File: lib/formats/arc/fjsys_archive.rb
require 'lib/formats/gfx/mgd_converter'
require 'lib/formats/script/msd_converter'
require 'lib/virtual_file'

# FJSYS archive
# Company: various
# Extension: none
# Known games:
# - Sono Hanabira ni Kuchizuke o 1
# - Sono Hanabira ni Kuchizuke o 2
# - Sono Hanabira ni Kuchizuke o 3
# - Sono Hanabira ni Kuchizuke o 4
# - Sono Hanabira ni Kuchizuke o 5
# - Sono Hanabira ni Kuchizuke o 6
# - Sono Hanabira ni Kuchizuke o 7
# - Sono Hanabira ni Kuchizuke o 8
# - Sono Hanabira ni Kuchizuke o 9
# - Sono Hanabira ni Kuchizuke o 10
# - Sono Hanabira ni Kuchizuke o 11
module FjsysArchive
  module_function

  MAGIC = "FJSYS\x00\x00\x00"

  def add_cli_help(arg_parser)
    MsdConverter.add_cli_help(arg_parser)
    MgdConverter.add_cli_help(arg_parser)
  end

  def parse_cli_options(arg_parser, options)
    MsdConverter.parse_cli_options(arg_parser, options)
    MgdConverter.parse_cli_options(arg_parser, options)
  end

  class Unpacker
    def unpack(arc_file, output_files, options)
      magic = arc_file.read(MAGIC.length)
      fail RecognitionError, 'Not a FJSYS archive' unless magic == MAGIC

      header_size,
      file_names_size,
      file_count = arc_file.read(76).unpack('LLLx64')
      file_names_start = header_size - file_names_size

      file_count.times do
        output_files.write do
          file_name_origin,
          data_size,
          data_origin = arc_file.read(16).unpack('LLQ')

          file_name = arc_file.peek(file_name_origin + file_names_start) do
            arc_file.read_until_zero
          end

          data = arc_file.peek(data_origin) { arc_file.read(data_size) }

          file = VirtualFile.new(file_name, data)
          decode!(file, options)
          file
        end
      end
    end

    private

    def decode!(file, options)
      if file.data.start_with?(MgdConverter::MAGIC)
        MgdConverter.decode!(file, options)
      elsif file.name.downcase.end_with?('.msd')
        MsdConverter.decode!(file, options)
      end
    end
  end

  class Packer
    def pack(arc_file, input_files, _options)
      arc_file.write(MAGIC)

      file_names_start = input_files.length * 16 + 0x54
      file_names_size = input_files.names.map { |f| f.length + 1 }.reduce(0, :+)

      header_size = file_names_size + file_names_start

      arc_file.write([
        header_size,
        file_names_size,
        input_files.length].pack('LLLx64'))
      arc_file.write("\x00" * (header_size - arc_file.tell))

      table_entries = []
      input_files.each do |file|
        table_entries.push(
          file_name: file.name,
          data_size: file.data.length,
          data_origin: arc_file.tell)

        arc_file.write(file.data)
      end

      table_entries = fix_file_order(table_entries)

      arc_file.seek(file_names_start)
      table_entries.each do |e|
        e[:file_name_origin] = arc_file.tell - file_names_start
        arc_file.write(e[:file_name])
        arc_file.write("\x00")
      end

      arc_file.seek(0x54)
      table_entries.each do |e|
        arc_file.write([
          e[:file_name_origin],
          e[:data_size],
          e[:data_origin]].pack('LLQ'))
      end
    end

    private

    # it is important to sort the files like the game did,
    # because the game refers to the file by their indices, not the file names!
    def fix_file_order(table_entries)
      # I guess the files are sorted alphabetically, but this assumption might
      # be wrong. I know that what I do below is stupid, but "_" needed to be
      # placed before "0" and after "." in the games I tested.
      table_entries.sort_by { |e| e[:file_name].gsub('_', '/').downcase }
    end
  end
end

File: src/formats/arc/fjsys_archive.cc
// FJSYS archive
//
// Company:   various
// Engine:    -
// Extension: none
//
// Known games:
// - Sono Hanabira ni Kuchizuke o 1
// - Sono Hanabira ni Kuchizuke o 2
// - Sono Hanabira ni Kuchizuke o 3
// - Sono Hanabira ni Kuchizuke o 4
// - Sono Hanabira ni Kuchizuke o 5
// - Sono Hanabira ni Kuchizuke o 6
// - Sono Hanabira ni Kuchizuke o 7
// - Sono Hanabira ni Kuchizuke o 8
// - Sono Hanabira ni Kuchizuke o 9
// - Sono Hanabira ni Kuchizuke o 10
// - Sono Hanabira ni Kuchizuke o 11

#include "formats/arc/fjsys_archive.h"
#include "formats/gfx/mgd_converter.h"

namespace
{
    const std::string magic("FJSYS\x00\x00\x00", 8);

    typedef struct
    {
        size_t header_size;
        size_t file_names_size;
        size_t file_count;
    } Header;

    std::unique_ptr<Header> read_header(IO &arc_io)
    {
        std::unique_ptr<Header> header(new Header);
        header->header_size = arc_io.read_u32_le();
        header->file_names_size = arc_io.read_u32_le();
        header->file_count = arc_io.read_u32_le();
        arc_io.skip(64);
        return header;
    }

    std::unique_ptr<VirtualFile> read_file(
        IO &arc_io,
        const MgdConverter &mgd_converter,
        const Header &header)
    {
        std::unique_ptr<VirtualFile> file(new VirtualFile);
        size_t file_name_offset = arc_io.read_u32_le();
        size_t data_size = arc_io.read_u32_le();
        size_t data_offset = static_cast<size_t>(arc_io.read_u64_le());
        size_t old_pos = arc_io.tell();
        size_t file_names_start = header.header_size
            - header.file_names_size;

        arc_io.seek(file_name_offset + file_names_start);
        file->name = arc_io.read_until_zero();

        arc_io.seek(data_offset);
        file->io.write_from_io(arc_io, data_size);
        mgd_converter.try_decode(*file);

        arc_io.seek(old_pos);
        return file;
    }
}

struct FjsysArchive::Internals
{
    MgdConverter mgd_converter;
};

FjsysArchive::FjsysArchive() : internals(new Internals)
{
}

FjsysArchive::~FjsysArchive()
{
}

void FjsysArchive::add_cli_help(ArgParser &arg_parser) const
{
    internals->mgd_converter.add_cli_help(arg_parser);
}

void FjsysArchive::parse_cli_options(ArgParser &arg_parser)
{
    internals->mgd_converter.parse_cli_options(arg_parser);
}

void FjsysArchive::unpack_internal(
    VirtualFile &file, OutputFiles &output_files) const
{
    if (file.io.read(magic.size()) != magic)
        throw std::runtime_error("Not a FJSYS archive");

    std::unique_ptr<Header> header = read_header(file.io);

    for (size_t i = 0; i < header->file_count; i ++)
    {
        output_files.save([&]()
        {
            return read_file(file.io, internals->mgd_converter, *header);
        });
    }
}
